src/Form/PolicyForm.php

```diff
@@ -0,0 +1,173 @@
+      $container->get('entity_type.manager')->getStorage('user_role'),
+      $container->get('messenger')
+    );
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function form(array $form, FormStateInterface $form_state): array {
+    $form = parent::form($form, $form_state);
+
+    $form['#title'] = $this->t('Edit policy');
+
+    $form['name'] = [
+      '#type' => 'textfield',
+      '#title' => $this->t('Name'),
+      '#default_value' => $this->entity->getName(),
+      '#required' => TRUE,
+    ];
+
+    $options = [];
+    foreach ($this->roleStorage->loadMultiple() as $role) {
+      $id = $role->id();
+      if ($id == 'anonymous') {
+        continue;
+      }
+
+      $options[$id] = $role->label();
+    }
+
+    $form['role'] = [
+      '#type' => 'select',
+      '#options' => $options,
+      '#title' => $this->t('User role'),
+      '#description' => $this->t('Select a role for which this policy should be applied.<br>When the user is registering only <em>All</em> and <em>Authenticated user</em> role policies will be used.'),
+      '#default_value' => $this->entity->getRole(),
+      '#required' => TRUE,
+      '#disabled' => !$this->entity->isNew(),
+    ];
+
+    $form['minimumRequiredConstraints'] = [
+      '#type' => 'number',
+      '#title' => $this->t('Minimum required constraints'),
+      '#min' => 1,
+      '#required' => TRUE,
+      '#default_value' => $this->entity->getMinimumRequiredConstraints() ?? 1,
+    ];
+
+    $expire_days = $this->entity->getExpireDays();
+    $form['expire_password'] = [
+      '#type' => 'checkbox',
+      '#title' => $this->t('Expire password'),
+      '#default_value' => $expire_days > 0,
+    ];
+
+    $form['expireDays'] = [
+      '#type' => 'number',
+      '#title' => $this->t('Expire password after'),
+      '#field_suffix' => $this->t('day(s).'),
+      '#default_value' => $expire_days,
+      '#min' => 0,
+      '#states' => [
+        'visible' => [
+          ':input[name="expire_password"]' => ['checked' => TRUE],
+        ],
+      ],
+    ];
+
+    $form['priority'] = [
+      '#type' => 'number',
+      '#title' => $this->t('Priority'),
+      '#default_value' => $this->entity->getPriority() ?? 1,
+      '#min' => 0,
+      '#states' => [
+        'invisible' => [
+          ':input[name="role"]' => ['value' => AccountInterface::AUTHENTICATED_ROLE],
+        ],
+      ],
+    ];
+
+    return $form;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function validateForm(array &$form, FormStateInterface $form_state): void {
+    parent::validateForm($form, $form_state);
+
+    if ($this->entity->isNew()) {
+      $role = $form_state->getValue('role');
+      if (!empty($this->policyConfigEntityStorage->load($role))) {
+        $form_state->setError($form['role'], $this->t('A policy is already created for the %role role.', [
+          '%role' => $this->roleStorage->load($role)->label(),
```

Comment:
If you cache loaded roles in `form()` you do not need to load them again.

Reply:
Can you please explain it a bit more?

Reply:
In the form builder method, save the filtered roles to the form state storage so you do not need to load _all_ roles (instead of the filtered list of roles) here again.

Reply:
Okay, that makes sense, thanks!